jsk_pcl_ros/src/euclidean_cluster_extraction_nodelet.cpp

```diff
@@ -42,109 +42,270 @@ using namespace pcl;
 namespace jsk_pcl_ros
 {
 
-  void EuclideanClustering::extract(
-    const sensor_msgs::PointCloud2ConstPtr &input)
-  {
+  void EuclideanClustering::clusteringClusterIndices(
+    pcl::PointCloud<pcl::PointXYZ>::Ptr &cloud,
+    std::vector<pcl::PointIndices::Ptr> &cluster_indices,
+    std::vector<pcl::PointIndices> &clustered_indices) {
+
     boost::mutex::scoped_lock lock(mutex_);
-    vital_checker_->poke();
-    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud
-      (new pcl::PointCloud<pcl::PointXYZ>);
-    pcl::fromROSMsg(*input, *cloud);
-    
-    std::vector<pcl::PointIndices> cluster_indices;
-    // list up indices which are not NaN to use
-    // organized pointcloud
-    pcl::PointIndices::Ptr nonnan_indices (new pcl::PointIndices);
-    for (size_t i = 0; i < cloud->points.size(); i++) {
-      pcl::PointXYZ p = cloud->points[i];
-      if (!std::isnan(p.x) && !std::isnan(p.y) && !std::isnan(p.z)) {
-        nonnan_indices->indices.push_back(i);
+
+    clustered_indices.clear();
+
+    pcl::VoxelGrid<pcl::PointXYZ> voxel;
+    pcl::PointCloud<pcl::PointXYZ>::Ptr voxel_cloud(
+      new pcl::PointCloud<pcl::PointXYZ>);
+    pcl::PointCloud<pcl::PointXYZ>::Ptr preprocessed_cloud;
+    if (downsample_) {
+      voxel.setLeafSize(leaf_size_, leaf_size_, leaf_size_);
+      voxel.setSaveLeafLayout(true);
+      voxel.setInputCloud(cloud);
+      voxel.filter(*voxel_cloud);
+      preprocessed_cloud = voxel_cloud;
+      downsample_to_original_indices_.resize(cloud->points.size());
+      original_to_downsample_indices_.resize(cloud->points.size());
+      std::fill(original_to_downsample_indices_.begin(),
+                original_to_downsample_indices_.end(),
+                -1);
+      std::fill(downsample_to_original_indices_.begin(),
+                downsample_to_original_indices_.end(),
+                std::vector<int>());
+      for (size_t i = 0; i < cloud->points.size(); ++i) {
+        pcl::PointXYZ p = cloud->points[i];
+        if (std::isnan(p.x) || std::isnan(p.y) || std::isnan(p.z)) {
+          continue;
+        }
+        int index = voxel.getCentroidIndex(p);
+        if (index == -1) {
+          continue;
+        }
+        original_to_downsample_indices_[i] = index;
+        downsample_to_original_indices_[index].push_back(i);
```

Comment:
It may be possible to use the downsample function from the subclass, this line should be implemented as a class method.

Reply:
That's nice idea! Could you create a PR for this branch?

Reply:
ok, I created a PR to your branch

Reply:
Merged your PR. Thanks!
Please review it again.